libusb/os/windows_hotplug.c

```diff
@@ -0,0 +1,307 @@
+	for_each_device(ctx, dev)
+	{
+		((struct winusb_device_priv *)usbi_get_device_priv(dev))->hotplug_status = LEFT;
+	}
+
+	return ((struct windows_context_priv *)usbi_get_context_priv(ctx))->backend->get_device_list(ctx);
+}
+
+int windows_initial_scan_devices(struct libusb_context *ctx)
+{
+	usbi_mutex_static_lock(&active_contexts_lock);
+	
+	int ret =  windows_get_device_list(ctx);
+	
+	usbi_mutex_static_unlock(&active_contexts_lock);
+	return ret;
+}
+
+static void windows_refresh_device_list(struct libusb_context *ctx)
+{
+	int ret = windows_get_device_list(ctx);
+	if(ret != LIBUSB_SUCCESS)
+	{
+		usbi_err(ctx, "hotplug failed to retrieve current list with error: %s", libusb_error_name(ret)); 
+		return;
+	}
+
+	struct libusb_device *dev, *next_dev;
+	struct winusb_device_priv *priv;
+
+	for_each_device_safe(ctx, dev, next_dev)
+	{
+		priv = usbi_get_device_priv(dev);
+		if(priv->hotplug_status != LEFT)
+		{
+			continue;
+		}
+
+		if(priv->initialized)
+		{
+			usbi_disconnect_device(dev);			
+		}
+		else
+		{
+			usbi_detach_device(dev);	
+		}
+	}
+
+	for_each_device(ctx, dev)
+	{
+		priv = usbi_get_device_priv(dev);
+
+		if(priv->hotplug_status != ARRIVED)
+		{
+			continue;
+		}
+
+		usbi_hotplug_notification(ctx, dev, LIBUSB_HOTPLUG_EVENT_DEVICE_ARRIVED);
+	}
+}
+
+static void windows_refresh_device_list_for_all_ctx(void)
+{
+	usbi_mutex_static_lock(&active_contexts_lock);
+
+	struct libusb_context *ctx;
+	for_each_context(ctx)
+	{
+		windows_refresh_device_list(ctx);
+	}
+	usbi_mutex_static_unlock(&active_contexts_lock);
+}
+
+#define WND_CLASS_NAME TEXT("libusb-1.0-windows-hotplug")
+
+static bool init_wnd_class(void)
+{
+	WNDCLASS wndClass = { 0 };
+	wndClass.lpfnWndProc = windows_proc_callback;
+	wndClass.hInstance = GetModuleHandle(NULL);
+	wndClass.lpszClassName = WND_CLASS_NAME;
+
+	if (!RegisterClass(&wndClass)) 
+	{
+		log_error("event thread: RegisterClass");
+		return false;
+	}
+
+	return true;
+}
+
+static DWORD WINAPI windows_event_thread_main(LPVOID lpParam)
+{
+	UNUSED(lpParam);
+
+	usbi_dbg(NULL, "windows event thread entering");
+
+	if (!init_wnd_class())
+	{
+		return (DWORD)-1;
+	}
+
+	windows_event_hWnd = CreateWindow(
+		WND_CLASS_NAME,
+		TEXT(""),
+		0, 
+		0, 0, 0, 0,
+		NULL, NULL,
+		GetModuleHandle(NULL),
+		NULL);
+
+	if (windows_event_hWnd == NULL)
+	{
+		log_error("event thread: CreateWindow");
+		return (DWORD)-1;
+	}
+
+	MSG msg;
+	BOOL ret_val;
+
+	while ((ret_val = GetMessage(&msg, windows_event_hWnd, 0, 0)) != 0)
+	{
+		if (ret_val == -1)
+		{
+			log_error("event thread: GetMessage");
+			break;
+		}
+
+		if(!SUCCEEDED(TranslateMessage(&msg)))
+		{
+			log_error("event thread: TranslateMessage");
+		}
+
+		if(!SUCCEEDED(DispatchMessage(&msg)))
+		{
+			log_error("event thread: DispatchMessage");
+		}
+	}
+
+	usbi_dbg(NULL, "windows event thread exiting");
+
+	return 0;
+}
+
+static bool register_device_interface_to_window_handle(
+	IN GUID interface_class_guid,
+	IN HWND hWnd,
```

Comment:
I guess the hWnd name is always used in such code by convention? Otherwise it would be a bit cryptic for the uninitiated :)

Reply:
actually I checked and the most common capitalization is `hwnd`, I will change

Reply:
There is really both hWnd and hwnd found in the wild, see here for instance: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getmessage
So it is a matter of taste, I guess.
I'll change to hwnd anyway.